kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+    `analyze` request while another analysis is pending.
+    
+    The reply includes the input file(s) on which to begin analysis, as well as
+    the output file to which to emit analysis output. It reflects some of the
+    kythe.proto.CompilationUnit structure. The driver (and driver environment)
+    MUST make contents available for each file referenced in the compilation
+    record at the file system path specified in the compilation record.
+
+    The output file should contain the content appropriate to the
+    analysis-type, encoded as specified in the `InitRequest`. For Kythe index
+    analyses, this would be `kythe.common.Entry` objects. Other analyzers might
+    emit a "findings" structure.
+
+    Note that contents need not be literally copied. An overlay filesystem,
+    hardlinks, symlinks, or other may be used.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":2,"method":"analyze","params":{"types":["kythe/indexer/go"]}}
+        « {"jsonrpc":"2.0","id":2","result":{"input":"path/to/main.go", "output":"path/to/output.json"}}
+
+-   `vname(VNameRequest)`
+
+    Requests the complete VName to emit for the specified path and signature.
+
+-   `log(LogRequest)`
+
+    A notification from the analyzer to the environment, containing a message
+    string to be logged in association with the current analysis. The
+    environment does not reply.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"log","params":{"message":{...}}}
+
+-   `done(LogRequest)`
+
+    A notification from the analyzer to the environment, signifying that the
+    analyzer has completed work on the specified analysis ID. The environment
+    does not reply. After receiving this notification, the driver MUST discard
+    any further requests for the analysis ID (and return an error, if
+    appropriate) unless and until it issues a new analysis reusing that ID.
+
+    Example:
+
+        » {"jsonrpc":"2.0","method":"done","params":{"message":"success"}}
+
+### Errors
+
+In addition to the standard errors defined by JSON-RPC, this protocol defines
+these errors:
+
+-   `ProtocolError` (code: -1). Indicates a violation of the protocol.
+
+## Input filesystem
+
+This RFC essentially postulates that the driver presents the analyzer
+a *limited* virtualization of a filesystem, specific to each
+analysis. In the context of a single analysis:
+
+-   For each file path occurring in the compilation record, the driver MUST 
+    support `open()`, `read()` and `lseek()` (or equivalent) operations. The
+    contents backing such operations MUST be precisely the contents specified
+    in the compilation record. Other file operations MAY be supported.
+
+-   For each directory implied by the file paths in the compilation record, the
+    driver MAY support a subset of read-only directory operations, e.g. in order
+    to provide directory listings via `readdir(3)` or equivalent.
+
+-   For all other paths, the driver MAY fall back to the underlying host
```

Comment:
I may be misunderstanding: does this mean to imply that paths in the compilation record MUST NOT fall back to the underlying host? Or is it just saying that access to the filesystem is not necessarily limited to compilation record paths?

Reply:
The latter. Does the "MAY" not convey that sense sufficiently? I'm open to other wording.

Reply:
I am perhaps being overly literal and getting myself confused. Since falling back to the underlying host is not mentioned above, one might infer it is not allowed. It may help to replace "For all other paths..." with "For paths that are available via the host's local filesystem..."